### PR TITLE
Dataset shuffle

### PR DESCRIPTION
**Pull request by orm011**
Opening for comments.

- - -
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/49)

### CODE REVIEW

File: pixeltable/dataframe.py [hunk 1]
```diff
@@ -580,13 +580,19 @@ def to_pytorch_dataset(self, image_format : str = 'pt') -> 'torch.utils.data.Ite
         from pixeltable.utils.parquet import save_parquet # pylint: disable=import-outside-toplevel
         from pixeltable.utils.pytorch import PixeltablePytorchDataset # pylint: disable=import-outside-toplevel
 
-        summary_string = json.dumps(self._as_dict()) 
+        if shuffle:
+            assert self.order_by_clause is None or self.order_by_clause == [], 'order_by() cannot be used with shuffle=True'
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: we should never have an empty order by clause

File: pixeltable/tests/test_dataframe.py [hunk 2]
```diff
@@ -285,7 +285,8 @@ def restrict_json_for_default_collate(obj):
             c_image=t.c_image.resize([220, 224]).convert('RGB')
         )
         df_size = df.count()
-        ds = df.to_pytorch_dataset(image_format='pt')
+        assert df_size >= 23, 'want to test enough tuples'
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: 23, huh?

File: pixeltable/tests/test_dataframe.py [hunk 3]
```diff
@@ -295,18 +296,26 @@ def restrict_json_for_default_collate(obj):
         # test we get all rows
         def check_recover_all_rows(ds, size : int, **kwargs):
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py:365]: return type missing

File: pixeltable/tests/test_dataframe.py [hunk 4]
```diff
@@ -285,7 +285,8 @@ def restrict_json_for_default_collate(obj):
             c_image=t.c_image.resize([220, 224]).convert('RGB')
         )
         df_size = df.count()
-        ds = df.to_pytorch_dataset(image_format='pt')
+        assert df_size >= 23, 'want to test enough tuples'
+        ds = df.to_pytorch_dataset(image_format='pt')
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: since you're explicitly testing the shuffling, you should supply the parameter explicitly as well

File: pixeltable/tests/test_dataframe.py [hunk 5]
```diff
@@ -295,18 +296,26 @@ def restrict_json_for_default_collate(obj):
...
-            assert loaded_ids == set(range(size))
+            assert set(loaded_ids) == set(range(size))
+            return loaded_ids
+
+        loaded_ids = check_recover_all_rows(ds, size=df_size, batch_size=3, num_workers=0) # within this process
+        assert loaded_ids != list(range(df_size)) # check ordering is not preserved
+
+        ds_notshuffled = df.to_pytorch_dataset(image_format='pt', shuffle=False)
+        loaded_ids = check_recover_all_rows(ds_notshuffled, size=df_size, batch_size=3, num_workers=0)
+        assert loaded_ids == list(range(df_size)) # check ordering is preserved
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: This isn't guaranteed to work, because the ordering might change, depending on what Postgres decides. In this case, you should supply an explicit order_by().

File: pixeltable/tests/test_dataframe.py [hunk 6]
```diff
@@ -295,18 +296,26 @@ def restrict_json_for_default_collate(obj):
...
 
         # check different number of workers
-        check_recover_all_rows(ds, size=df_size, batch_size=3, num_workers=0) # within this process
-        check_recover_all_rows(ds, size=df_size, batch_size=3, num_workers=2) # two separate processes
+        loaded_ids = check_recover_all_rows(ds, size=df_size, batch_size=3, num_workers=2) # two separate processes
+        assert loaded_ids != list(range(df_size)) # check ordering is not preserved
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: missing: check for failure of order_by() + shuffle

**Review by mkornacker — CHANGES_REQUESTED** [review]

File: pixeltable/tests/test_dataframe.py (hunk 4 above)
Comment (orm011) [reply pixeltable/tests/test_dataframe.py]: I'm also testing that the default argument is indeed shuffle, which we wouldn't notice if changed if we explicitly test positive.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 584 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421825348:-NoTtB1eB_P9Sd3NhYR0:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/dataframe.py#L584)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

we should never have an empty order by clause
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_dataframe.py` line 288 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421826540:-NoTzWpY92MIhtcg6m8a:b-7h91p9) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L288)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

23, huh?
</blockquote></details>

added comment. prime number plus large enough.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/tests/test_dataframe.py` line 297 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421826737:-NoTzcyI8RM4gjaxKX6c:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L297)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

return type missing
</blockquote></details>

Done.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/tests/test_dataframe.py` line 314 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421828408:-NoTzeTu3m7at-FI_1on:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L314)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This isn't guaranteed to work, because the ordering might change, depending on what Postgres decides. In this case, you should supply an explicit order_by().
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_dataframe.py` line 318 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421828557:-NoTzfDtAdmMXFFbd78t:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L318)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

missing: check for failure of order_by() + shuffle
</blockquote></details>

Done.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/tests/test_dataframe.py` line 289 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421828092:-NoZZnI-9MqpC1LpvjIf:b3f5gaf) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L289)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

I'm also testing that the default argument is indeed shuffle, which we wouldn't notice if changed if we explicitly test positive.
</blockquote></details>

broke this test up into its own thing, to untangle dataloader stuff from dataset shuffle, which was complicating this dataloader test
 (see new test )
___
*[`pixeltable/tests/test_dataframe.py` line 314 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421828408:-NoZ_1az8WGsA1BlmHfT:b-cep6bg) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L314)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

see new test
___
*[`pixeltable/tests/test_dataframe.py` line 318 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/49#gh-1421828557:-NoZ_2i22yu9CFtDjGR6:b-cep6bg) ([raw file](https://github.com/mkornacker/pixeltable/blob/f0032e86f3ea58194bbdbb7703343a0bcf81e9b8/pixeltable/tests/test_dataframe.py#L318)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

see new test

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/tests/test_dataframe.py` line 238 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/49#-NoZ_6Oi-6ETC6yInRwK:-NoZ_6Oi-6ETC6yInRwL:b-adefgb) ([raw file](https://github.com/mkornacker/pixeltable/blob/885433e33bf5fc88d7696c682283785f3e151704/pixeltable/tests/test_dataframe.py#L238)):*
> ```Python
>             assert isinstance(tup['c_json'], dict)
> 
>     def test_to_pytorch_dataset_shuffle(self, all_datatypes_tbl: catalog.MutableTable) -> None:
> ```

new test to test shuffle separately, without needing a dataloader

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 8 unresolved discussions (waiting on @orm011)

___
*[`pixeltable/dataframe.py` line 630 at r4](https://reviewable.io/reviews/mkornacker/pixeltable/49#-NpNhuo90-6Cx8jSqrqk:-NpNhuo90-6Cx8jSqrql:b-xd5q8l) ([raw file](https://github.com/mkornacker/pixeltable/blob/8ad181548e15f1da362f40ec86c250b0b068be3f/pixeltable/dataframe.py#L630)):*
> ```Python
>             if not self.order_by_clause is None:
>                 raise exc.Error('order_by() cannot be used with shuffle')
>             df = self.order_by(exprs.SqlRandom(), asc=True)
> ```

Doesn't this mean that even if you call this with shuffle=True, you'll get the same order as last time when it's coming from the cache? I think that would lead to a lot of confusion and would probably also make the feature not be useful.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49)* status: 0 of 7 files reviewed, 8 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 630 at r4](https://reviewable.io/reviews/mkornacker/pixeltable/49#-NpNhuo90-6Cx8jSqrqk:-NpQItsL2_-SoS6Z_rUL:b-lx5c8y) ([raw file](https://github.com/mkornacker/pixeltable/blob/8ad181548e15f1da362f40ec86c250b0b068be3f/pixeltable/dataframe.py#L630)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Doesn't this mean that even if you call this with shuffle=True, you'll get the same order as last time when it's coming from the cache? I think that would lead to a lot of confusion and would probably also make the feature not be useful.
</blockquote></details>

Not sure I follow, but my interpretation of what you are asking is a scenario where 1.  a call is made without shuffle, then 2. a call is made with shuffle, the question being what's the cache behavior.

The reasonable behavior is for the output in 2. to be shuffled.
BC. the cache key is for`df`, and not for `self`, the intention is these two variations have different cached version, as would a separate 'order by '.

**Comment by mkornacker** [comment]
My point was: if you call to_dataset() twice in a row with shuffle=True, you'll get the same dataset back, with the exact same row order. I don't think this is what people would expect or want.

**Review by orm011 — COMMENTED** [review]
I can see why this may seem counter-intuitive, at the same time, I think if I were using it to train a model, its actually good to have a way to re-use it, mostly because the idea of shuffling is so that each mini-batch can be seen as providing an unbiased 
"poll" estimate of the Loss gradient. So, re-using the same shuffling is not bad, depending on what you are doing, eg if you want to try different experiments, re-using the same shuffling rather than redoing it every time is not bad.  The alternative, of forcing a re-shuffle seems less desirable.

One option is to add a shuffle seed parameter, and have be set to a default. If the user really cares for a different order, they can change the seed. This enables both scenarios, but I'd say the default seed is ok, if someone really cares to pay that extra price then they can go out of their way.  

> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/49#-:-NpRiwJH52kgg7YaNoAG:bkji00u)* status: 0 of 7 files reviewed, 8 unresolved discussions (waiting on @mkornacker)
